org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -292,10 +292,29 @@ public IStatus run(IProgressMonitor monitor) {
 					pm.projectsImported(monitor);
 
 					IndexUtils.copyIndexesToSharedLocation();
-					JobHelpers.waitForBuildJobs(60 * 60 * 1000); // 1 hour
-					logInfo(">> build jobs finished");
-					workspaceDiagnosticsHandler.publishDiagnostics(monitor);
-				} catch (OperationCanceledException | CoreException e) {
+					Job waitThenPublishDiagnostics = new Job("Wait for builds, then publish diagnostics") { // should not run in workspace!
+						@Override
+						public IStatus run(IProgressMonitor monitor) {
+							JobHelpers.waitForBuildJobs(60 * 60 * 1000); // 1 hour
+							logInfo(">> build jobs finished");
+							try {
+								workspaceDiagnosticsHandler.publishDiagnostics(monitor);
+							} catch (CoreException e) {
+								logException(e.getMessage(), e);
+								return e.getStatus();
+							}
+							return Status.OK_STATUS;
+						}
+
+						@Override
+						public boolean belongsTo(Object family) {
+							return JAVA_LSP_INITIALIZE_WORKSPACE.equals(family);
+						}
```

Comment:
Can we define the job just below `initializeWorkspace` for the sake of make the nested portion smaller ?